libraries/abstractions/pkcs11/utest/iot_pkcs11_mbedtls_utest.c

```diff
@@ -171,112 +299,540 @@ void test_pkcs11_C_InitializeInitTwice( void )
+    xResult = prvInitializePkcs11( ( SemaphoreHandle_t ) &xResult );
 
     TEST_ASSERT_EQUAL( CKR_OK, xResult );
 
     xResult = C_Initialize( NULL );
     TEST_ASSERT_EQUAL( CKR_CRYPTOKI_ALREADY_INITIALIZED, xResult );
 
+    xResult = prvUninitializePkcs11();
+    TEST_ASSERT_EQUAL( CKR_OK, xResult );
+}
+
+/* ======================  TESTING C_Finalize  ============================ */
+
+/*!
+ * @brief C_Finalize initialize was not already called.
+ *
+ */
+void test_pkcs11_C_FinalizeUninitialized( void )
+{
+    CK_RV xResult = CKR_OK;
+
+    xResult = prvUninitializePkcs11();
+    TEST_ASSERT_EQUAL( CKR_CRYPTOKI_NOT_INITIALIZED, xResult );
+}
+
+/*!
+ * @brief C_Finalize was called with a non-null pointer.
+ *
+ */
+void test_pkcs11_C_FinalizeBadArgs( void )
+{
+    CK_RV xResult = CKR_OK;
+
     mbedtls_entropy_free_CMockIgnore();
     mbedtls_ctr_drbg_free_CMockIgnore();
     vQueueDelete_CMockIgnore();
-    xResult = C_Finalize( NULL );
+    xResult = C_Finalize( ( CK_VOID_PTR ) &xResult );
+
+    TEST_ASSERT_EQUAL( CKR_ARGUMENTS_BAD, xResult );
+}
+
+/* ======================  TESTING C_GetFunctionList  ============================ */
+
+/*!
+ * @brief C_GetFunctionList happy path.
+ *
+ */
+void test_pkcs11_C_GetFunctionList( void )
+{
+    CK_RV xResult = CKR_OK;
+    CK_FUNCTION_LIST_PTR pxFunctionList = NULL;
+
+    xResult = C_GetFunctionList( &pxFunctionList );
 
     TEST_ASSERT_EQUAL( CKR_OK, xResult );
 }
 
-/* ======================  TESTING C_Finalize  ============================ */
+/*!
+ * @brief C_GetFunctionList bad args.
+ *
+ */
+void test_pkcs11_C_GetFunctionListBadArgs( void )
+{
+    CK_RV xResult = CKR_OK;
+
+    xResult = C_GetFunctionList( NULL );
+
+    TEST_ASSERT_EQUAL( CKR_ARGUMENTS_BAD, xResult );
+}
+
+/* ======================  TESTING C_GetSlotList  ============================ */
 
 /*!
- * @brief C_Finalize happy path.
+ * @brief C_GetSlotList happy path.
  *
  */
-void test_pkcs11_C_Finalize( void )
+void test_pkcs11_C_GetSlotList( void )
 {
     CK_RV xResult = CKR_OK;
+    CK_SLOT_ID_PTR pxSlotId = NULL;
+    CK_ULONG xSlotCount = 0;
+    CK_BBOOL xTokenPresent = CK_TRUE;
 
-    xQueueCreateMutex_IgnoreAndReturn( ( SemaphoreHandle_t ) &xResult );
-    CRYPTO_Init_Ignore();
-    mbedtls_entropy_init_Ignore();
-    mbedtls_ctr_drbg_init_Ignore();
-    mbedtls_ctr_drbg_seed_IgnoreAndReturn( 0 );
-    xResult = C_Initialize( NULL );
+    xResult = prvInitializePkcs11( ( SemaphoreHandle_t ) &xResult );
+    TEST_ASSERT_EQUAL( CKR_OK, xResult );
 
+    xResult = C_GetSlotList( xTokenPresent, pxSlotId, &xSlotCount );
     TEST_ASSERT_EQUAL( CKR_OK, xResult );
+    TEST_ASSERT_EQUAL( 1, xSlotCount );
 
-    mbedtls_entropy_free_CMockIgnore();
-    mbedtls_ctr_drbg_free_CMockIgnore();
-    vQueueDelete_CMockIgnore();
-    xResult = C_Finalize( NULL );
+    pxSlotId = pvPkcs11MallocCb( sizeof( CK_SLOT_ID ) * xSlotCount, 1 );
+    memset( pxSlotId, 0, sizeof( CK_SLOT_ID ) * 1 );
```

Comment:
Why is the malloc done in such a way that there could be more than one slot, but the memset isn't?

Reply:
Done